File: NLIP.iShare.AuthorizationRegistry.Core/DelegationTranslateService.cs
using Newtonsoft.Json;
using NLIP.iShare.AuthorizationRegistry.Core.Api;
using NLIP.iShare.Models.DelegationEvidence;
using NLIP.iShare.Models.DelegationMask;
using System.Collections.Generic;
using System.Linq;

namespace NLIP.iShare.AuthorizationRegistry.Core
{
    public class DelegationTranslateService : IDelegationTranslateService
    {
        public DelegationTranslationTestResponse Translate(DelegationMask delegationMask, string policy)
        {
            var delegationEvidence = JsonConvert.DeserializeObject<DelegationTranslationTestResponse>(policy);

            var delegationResponse = PrepareResponse(delegationEvidence);

            foreach (var maskPolicySet in delegationMask.DelegationRequest.PolicySets)
            {
                foreach (var evidencePolicySet in delegationEvidence.DelegationEvidence.PolicySets)
                {
                    var responsePolicySet = GetResponsePolicySet(maskPolicySet, evidencePolicySet);
                    delegationResponse.DelegationEvidence.PolicySets.Add(responsePolicySet);
                }
            }
            return delegationResponse;
        }

        private static PolicySet GetResponsePolicySet(DelegationRequestPolicySet maskPolicySet, PolicySet evidencePolicySet)
        {
            var responsePolicySet = new PolicySet
            {
                MaxDelegationDepth = evidencePolicySet.MaxDelegationDepth,
                Target = evidencePolicySet.Target,
                Policies = new List<Policy>()
            };

            foreach (var maskPolicy in maskPolicySet.Policies)
            {
                var matchingPolicies = evidencePolicySet.Policies.Where(e => IsMatchingPolicy(maskPolicy, e));

                var responsePolicy = new Policy
                {
                    Target = maskPolicy.Target,
                    Rules = new List<PolicyRule>()
                };

                var rule = PolicyRule.Permit();

                if (!matchingPolicies.Any() || matchingPolicies.Any(e => AccessDeniedToContainers(maskPolicy, e)))
                {
                    rule = PolicyRule.Deny();
                }

                AddRuleAndPolicy(responsePolicySet, responsePolicy, rule);
            }

            return responsePolicySet;
        }

        private static void AddRuleAndPolicy(PolicySet responsePolicySet, Policy responsePolicy, PolicyRule rule)
        {
            responsePolicy.Rules.Add(rule);
            responsePolicySet.Policies.Add(responsePolicy);
        }

        private static DelegationTranslationTestResponse PrepareResponse(DelegationTranslationTestResponse delegationEvidence)
        {
            return new DelegationTranslationTestResponse
            {
                DelegationEvidence = new DelegationEvidence
                {
                    NotBefore = delegationEvidence.DelegationEvidence.NotBefore,
                    NotOnOrAfter = delegationEvidence.DelegationEvidence.NotOnOrAfter,
                    PolicyIssuer = delegationEvidence.DelegationEvidence.PolicyIssuer,
                    Target = delegationEvidence.DelegationEvidence.Target,
                    PolicySets = new List<PolicySet>()
                }
            };
        }

        private static bool IsMatchingPolicy(Policy maskPolicy, Policy evidencePolicy)
            => IsTypeMatch(maskPolicy, evidencePolicy)
               && HasAllIdentifiers(maskPolicy, evidencePolicy)
               && HasAllAttributes(maskPolicy, evidencePolicy)
               && HasAllActions(maskPolicy, evidencePolicy);

        private static bool IsTypeMatch(Policy maskPolicy, Policy evidencePolicy)
        {
            var resource = evidencePolicy.Target.Resource;

            var isTypeMatch = maskPolicy.Target.Resource.Type == resource.Type;
            return isTypeMatch;
        }

        private static bool HasAllIdentifiers(Policy maskPolicy, Policy evidencePolicy)
        {
            var resource = evidencePolicy.Target.Resource;
            return maskPolicy.Target.Resource.Identifiers.All(mId =>
                resource.Identifiers.Count == 1 &&
                resource.Identifiers.Any(eId => eId == "*") ||
                resource.Identifiers.Contains(mId));
        }

        private static bool HasAllAttributes(Policy maskPolicy, Policy evidencePolicy)
        {
            var resource = evidencePolicy.Target.Resource;
            return maskPolicy.Target.Resource.Attributes.All(mAtt =>
                resource.Attributes.Count == 1 &&
                resource.Attributes.Any(eAtt => eAtt == "*") ||
                resource.Attributes.Contains(mAtt));
        }

        private static bool HasAllActions(Policy maskPolicy, Policy evidencePolicy)
        {
            return maskPolicy.Target.Actions != null &&
                   maskPolicy.Target.Actions.Any() &&
                   maskPolicy.Target.Actions.All(mAct =>
                       evidencePolicy.Target.Actions.Count == 1 &&
                       evidencePolicy.Target.Actions.Any(eAct => eAct == "*") ||
                       evidencePolicy.Target.Actions.Contains(mAct));
        }

        private static bool AccessDeniedToContainers(Policy maskPolicy, Policy evidencePolicy)
        {
            var reversed = evidencePolicy.Rules.ToList();
            reversed.Reverse();

            var maskResource = maskPolicy.Target.Resource;

            return reversed.Any(rule =>
                rule.Effect == PolicyRule.Deny().Effect &&
                maskResource.Type == rule.Target.Resource.Type &&
                maskResource.Identifiers.Any(mId =>
                    rule.Target.Resource.Identifiers.Contains("*") ||
                    rule.Target.Resource.Identifiers.Contains(mId)) &&
                maskResource.Attributes.Any(mAtt =>
                    rule.Target.Resource.Attributes.Contains("*") ||
                    rule.Target.Resource.Attributes.Contains(mAtt))
            );
        }
    }
}

File: modern/authorization-registry/src/delegation/delegation-translate.service.ts
import { Injectable } from '@nestjs/common';
import {
  DelegationEvidence,
  DelegationEvidenceEnvelope,
  Policy,
  PolicyRule,
  PolicySet,
} from './dto/delegation-evidence.dto';
import {
  DelegationMask,
  DelegationRequestPolicySet,
} from './dto/delegation-mask.dto';

/**
 * Port of iSHARE.IdentityServer.Delegation.DelegationTranslateService.
 *
 * Given a stored policy (as JSON, the canonical iSHARE delegation evidence
 * envelope) and a delegation mask, derive a fresh envelope where each mask
 * policy is converted to a single Permit/Deny rule based on whether it is
 * fully covered by the stored policy.
 */
@Injectable()
export class DelegationTranslateService {
  translate(mask: DelegationMask, storedPolicyJson: string): DelegationEvidenceEnvelope {
    const stored = JSON.parse(storedPolicyJson) as DelegationEvidenceEnvelope;
    const out = prepareResponse(stored.delegationEvidence);

    for (const maskSet of mask.delegationRequest.policySets) {
      for (const storedSet of stored.delegationEvidence.policySets) {
        out.delegationEvidence.policySets.push(buildResponsePolicySet(maskSet, storedSet));
      }
    }

    return out;
  }
}

function prepareResponse(src: DelegationEvidence): DelegationEvidenceEnvelope {
  return {
    delegationEvidence: {
      notBefore: src.notBefore,
      notOnOrAfter: src.notOnOrAfter,
      policyIssuer: src.policyIssuer,
      target: src.target,
      policySets: [],
    },
  };
}

function buildResponsePolicySet(
  maskSet: DelegationRequestPolicySet,
  storedSet: PolicySet,
): PolicySet {
  const responseSet: PolicySet = {
    maxDelegationDepth: storedSet.maxDelegationDepth,
    target: storedSet.target,
    policies: [],
  };

  for (const maskPolicy of maskSet.policies) {
    const matching = storedSet.policies.filter((p) => isMatchingPolicy(maskPolicy, p));

    const rule: PolicyRule =
      matching.length === 0 || matching.some((p) => accessDeniedToContainers(maskPolicy, p))
        ? { effect: 'Deny' }
        : { effect: 'Permit' };

    responseSet.policies.push({ target: maskPolicy.target, rules: [rule] });
  }

  return responseSet;
}

function isMatchingPolicy(maskPolicy: Policy, stored: Policy): boolean {
  return (
    isTypeMatch(maskPolicy, stored) &&
    isProviderMatch(maskPolicy, stored) &&
    hasAllIdentifiers(maskPolicy, stored) &&
    hasAllAttributes(maskPolicy, stored) &&
    hasAllActions(maskPolicy, stored)
  );
}

function isTypeMatch(maskPolicy: Policy, stored: Policy): boolean {
  return maskPolicy.target.resource.type === stored.target.resource.type;
}

function hasAllIdentifiers(maskPolicy: Policy, stored: Policy): boolean {
  const storedIds = stored.target.resource.identifiers;
  return maskPolicy.target.resource.identifiers.every(
    (mId) => (storedIds.length === 1 && storedIds.includes('*')) || storedIds.includes(mId),
  );
}

function hasAllAttributes(maskPolicy: Policy, stored: Policy): boolean {
  const storedAttrs = stored.target.resource.attributes;
  return maskPolicy.target.resource.attributes.every(
    (mAtt) => (storedAttrs.length === 1 && storedAttrs.includes('*')) || storedAttrs.includes(mAtt),
  );
}

function hasAllActions(maskPolicy: Policy, stored: Policy): boolean {
  const maskActions = maskPolicy.target.actions ?? [];
  const storedActions = stored.target.actions ?? [];
  if (maskActions.length === 0) return false;
  return maskActions.every(
    (mAct) =>
      (storedActions.length === 1 && storedActions.includes('*')) || storedActions.includes(mAct),
  );
}

function isProviderMatch(maskPolicy: Policy, stored: Policy): boolean {
  const storedEnv = stored.target.environment;
  if (!storedEnv) return true;
  const maskEnv = maskPolicy.target.environment;
  if (!maskEnv) return false;
  return maskEnv.serviceProviders.every((sp) => storedEnv.serviceProviders.includes(sp));
}

/**
 * iSHARE rule: even if a Permit rule covers the request, a later Deny rule
 * with overlapping resource/action terms blocks the permit. The legacy
 * implementation iterates the rule list reversed, but logically it just
 * checks "any Deny rule that overlaps".
 */
function accessDeniedToContainers(maskPolicy: Policy, stored: Policy): boolean {
  return stored.rules.some((rule) => {
    if (rule.effect !== 'Deny' || !rule.target?.resource) return false;
    const r = rule.target.resource;
    const m = maskPolicy.target.resource;

    if (r.type !== m.type) return false;
    if (!hasAnyOverlap(r.identifiers, m.identifiers)) return false;
    if (r.attributes.length > 0 && !hasAnyOverlap(r.attributes, m.attributes)) return false;

    const maskActions = maskPolicy.target.actions ?? [];
    const ruleActions = rule.target.actions ?? [];
    if (ruleActions.length > 0 && !hasAnyOverlap(ruleActions, maskActions)) return false;

    return true;
  });
}

function hasAnyOverlap(a: string[], b: string[]): boolean {
  if (a.length === 0 || b.length === 0) return true; // legacy "permit when unset"
  return a.some((x) => b.includes(x));
}

function _silence(_p: PolicyRule) {
  // keep PolicyRule referenced even if narrowed away
}
void _silence;
